nipap/nipap/authlib.py

```diff
@@ -342,6 +343,104 @@ def authenticate(self):
+    """
+
+    _ldap_uri = None
+    _ldap_basedn = None
+    _ad_domain = None
+    _ad_rwgroup = None
+    _ad_rogroup = None
+    _ldap_conn = None
+    _authenticated = None
+
+    def __init__(self, name, username, password, authoritative_source, auth_options={}):
+        """ Constructor.
+
+            Note that the instance variables not are set by the constructor but
+            by the :func:`authenticate` method. Therefore, run the
+            :func:`authenticate`-method before trying to access those
+            variables!
+
+            * `name` [string]
+                Name of auth backend.
+            * `username` [string]
+                Username to authenticate as.
+            * `password` [string]
+                Password to authenticate with.
+            * `authoritative_source` [string]
+                Authoritative source of the query.
+            * `auth_options` [dict]
+                A dict which, if authenticated as a trusted user, can override
+                `username` and `authoritative_source`.
+        """
+
+
+        BaseAuth.__init__(self, username, password, authoritative_source, name, auth_options)
+        self._ldap_uri = self._cfg.get('auth.backends.' + self.auth_backend, 'uri')
+        self._ldap_basedn = self._cfg.get('auth.backends.' + self.auth_backend, 'basedn')
+        self._ad_domain = self._cfg.get('auth.backends.' + self.auth_backend, 'domain')
+        self._ad_rwgroup = self._cfg.get('auth.backends.' + self.auth_backend, 'rwgroup')
+        self._ad_rogroup = self._cfg.get('auth.backends.' + self.auth_backend, 'rogroup')
+
+        self._logger.debug('Creating AD Auth instance')
+
+        self._logger.debug('AD domain: ' + self._ad_domain)
+        self._logger.debug('AD LDAP URI: ' + self._ldap_uri)
+        self._ldap_conn = ldap.initialize(self._ldap_uri)
+
+
+
+    def authenticate(self):
+        """ Verify authentication.
+
+            Returns True/False dependant on whether the authentication
+            succeeded or not.
+        """
+
+        # if authentication has been performed, return last result
+        if self._authenticated is not None:
+            return self._authenticated
+
+        try:
+            self._ldap_conn.simple_bind_s(self.username + '@' + self._ad_domain, self.password)
+        except ldap.SERVER_DOWN as exc:
+            raise AuthError('Could not connect to LDAP server')
+        except (ldap.INVALID_CREDENTIALS, ldap.INVALID_DN_SYNTAX,
+                ldap.UNWILLING_TO_PERFORM) as exc:
+            # Auth failed
+            self._logger.debug('erroneous password for user %s' % self.username)
+            self._authenticated = False
+            return self._authenticated
+
+        # Get full name and group memberships
+        try:
+            res = self._ldap_conn.search_s(self._ldap_basedn, ldap.SCOPE_SUBTREE, '(sAMAccountName=' + self.username + ')',['cn','memberOf'])
```

Comment:
The value in self.username should be escaped (by calling ldap.filter.escape_filter_chars on it) before being included in the filter string passed to search_s.